src/integrations/authentication/implementations/openid/OpenID.ts

```diff
@@ -24,97 +39,121 @@ export default class OpenID implements IdentityProvider
 
     get connected(): boolean
     {
-        return this.#client !== undefined;
+        return this.#config !== undefined;
     }
 
     async connect(): Promise<void>
     {
-        const metaData: ClientMetadata =
-        {
-            client_id: this.#configuration.clientId,
-            client_secret: this.#configuration.clientSecret,
-            redirect_uris: [this.#configuration.redirectUri],
-            response_types: ['code']
-        };
+        const issuer = new URL(this.#configuration.issuer);
+        const clientId = this.#configuration.clientId;
+        const clientSecret = this.#configuration.clientSecret;
+        const requestOptions = this.#getRequestOptions();
 
-        const issuer = await Issuer.discover(this.#configuration.issuer);
-
-        this.#client = new issuer.Client(metaData);
+        this.#config = await discovery(issuer, clientId, clientSecret, undefined, requestOptions);
     }
 
     async disconnect(): Promise<void>
     {
-        this.#client = undefined;
+        this.#config = undefined;
     }
 
     async getLoginUrl(): Promise<string>
     {
-        const client = this.#getClient();
-        const parameters: AuthorizationParameters =
-        {
-            scope: 'openid profile email',
-            response_mode: 'form_post'
+        const redirect_uri = this.#configuration.redirectUri;
+        const scope = 'openid profile email';
+        const code_challenge = await calculatePKCECodeChallenge(this.#codeVerifier);
+        const code_challenge_method = 'S256';
+
+        const parameters: Record<string, string> = {
+            redirect_uri,
+            scope,
+            code_challenge,
+            code_challenge_method
         };
 
-        return client.authorizationUrl(parameters);
+        const config = this.#getConfig();
+        const redirectTo = buildAuthorizationUrl(config, parameters);
+
+        return redirectTo.href;
     }
 
     async login(data: Record<string, unknown>): Promise<Session>
     {
-        const client = this.#getClient();
-        const redirectUri = this.#configuration?.redirectUri;
+        const config = this.#getConfig();
+        const currentUrl = new URL(`${this.#configuration.redirectUri}?session_state=${data.session_state}&iss=${data.iss}&code=${data.code}`);
+
+        const tokens = await authorizationCodeGrant(config, currentUrl, {
+            pkceCodeVerifier: this.#codeVerifier,
+            idTokenExpected: true
+        });
+
+        const access_token = tokens.access_token as string;
+        const claims = tokens.claims();
 
-        const tokenSet = await client.callback(redirectUri, data);
-        const userinfo = await client.userinfo(tokenSet.access_token as string);
+        const sub = claims!.sub as string;
```

Comment:
According to the docs, the `claims` can be undefined when the auth server doesn't return a `expires_in` value. Shouldn't we check this?